File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.cpp
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = (int)heights.size();
        int ans = 0;

        stack<pair<int,int>> st;

        for (int i = 0; i < n; ++i) {
            int start = i;

            while (!st.empty() && st.top().second > heights[i]) {
                auto [j, h] = st.top();
                st.pop();
                ans = max(ans, h * (i - j));
                start = j;  
            }

            st.emplace(start, heights[i]);
        }

        while (!st.empty()) {
            auto [j, h] = st.top();
            st.pop();
            ans = max(ans, h * (n - j));
        }

        return ans;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)
        l = 0
        cnt = [0] * 26
        ans = 0
        for r ,x in enumerate(s):
            cnt[ord(x) - ord("A")] += 1
            while ( (r - l + 1) - max(cnt)  ) > k:
                cnt[ord(s[l]) - ord("A")] -= 1
                l += 1 
            ans = max(ans, r- l + 1)
        return ans
